packages/sdk/app-framework/src/react/Surface.tsx

```diff
@@ -9,13 +9,29 @@ import { byDisposition } from '@dxos/util';
 
 import { ErrorBoundary } from './ErrorBoundary';
 import { useCapabilities } from './useCapabilities';
-import { Capabilities, type SurfaceProps } from '../common';
+import { Capabilities, type SurfaceDefinition, type SurfaceProps } from '../common';
+import { type PluginsContext } from '../core';
 
 const useSurfaces = () => {
   const surfaces = useCapabilities(Capabilities.ReactSurface);
   return useMemo(() => surfaces.flat(), [surfaces]);
 };
 
+const findCandidates = (surfaces: SurfaceDefinition[], { role, data }: Pick<SurfaceProps, 'role' | 'data'>) => {
+  return Object.values(surfaces)
+    .filter((definition) =>
+      Array.isArray(definition.role) ? definition.role.includes(role) : definition.role === role,
+    )
+    .filter(({ filter }) => (filter ? filter(data ?? {}) : true))
+    .toSorted(byDisposition);
+};
+
+export const isSurfaceAvailable = (context: PluginsContext, { role, data }: Pick<SurfaceProps, 'role' | 'data'>) => {
```

Comment:
```suggestion
/**
 * @returns `true` if there is a contributed surface which matches the specified role & data, `false` otherwise.
 */
export const isSurfaceAvailable = (context: PluginsContext, { role, data }: Pick<SurfaceProps, 'role' | 'data'>) => {
```